### ISSUE TITLE
org.hibernate.MappingException: An association refers to an unmapped class

### ISSUE THREAD

**Issue by Euklios** [hank-cp/sbp#11]
So, I'm trying to understand how to use the "Single shared DataSource", as described in the [persistence documentation](https://github.com/hank-cp/sbp/blob/master/docs/persistence.md).

I have two plugins, one responsible for storing libraries and a second one for storing videos.

I'm using the SharedDataSourceSpringBootstrap within the SpringBootPlugin class in both plugins like so:
```java
public class LibraryPlugin extends SpringBootPlugin {
  public LibraryPlugin(final PluginWrapper wrapper) {
    super(wrapper);
  }

  @Override
  protected SpringBootstrap createSpringBootstrap() {
    return new SharedDataSourceSpringBootstrap(this, LibraryStarter.class);
  }
}
```

And create the Entities as usual, however, separated into the two different plugins.

The Video entity contains a reference to the Library entity as follows:
```
@Entity
public class Video {
  @Id
  private long id;
  @OneToOne
  @MapsId
  @JoinColumn(name = "library")
  private Library library;
}
```

When starting the system, the video-plugins exits with the following cause:
`Caused by: org.hibernate.MappingException: An association from the table video refers to an unmapped class: com.example.plugins.library.Library`

Now, as hibernate thinks that the plugin is unmapped and both plugins are running their own hibernate instance, where they can't see the other classes.

Is this even supposed to work?
In the end, the persistence documentation never states that the data source is also shared between plugins (i just assumed since both plugins are using the data source from the main-application)

Do I have to include SharedDataSourceSpringBootstrap somewhere within the main-application?
I have not seen anything that hints to something like that, but as the data-source is probably still separated.

Edit:
Forgot to mention, but i've added the plugin dependency to both, maven and plugin.properties

**Comment by Euklios** [comment]
By the way: The link to SharedDataSourceSpringBootstrap.java within the persistence-documentation is invalid (the "laxture" package is missing).

It currently is: 
https://github.com/hank-cp/sbp/blob/master/sbp-core/src/main/java/org/sbp/spring/boot/SharedDataSourceSpringBootstrap.java

But should be:
https://github.com/hank-cp/sbp/blob/master/sbp-core/src/main/java/org/laxture/sbp/spring/boot/SharedDataSourceSpringBootstrap.java

**Comment by hank-cp** [comment]
How you expose `LIbrary` eneity to video plugin?

**Comment by hank-cp** [comment]
I am afraid this issue may not be going to be fixed. Managing hibernate mapping graph is very very complicated and difficult (we are talking about a 20 years old monster here). 

So in the document, I do mention the `TransactionManager` could be shared, but not the Hibernate/JPA entity context.

**Comment by hank-cp** [comment]
> By the way: The link to SharedDataSourceSpringBootstrap.java within the persistence-documentation is invalid (the "laxture" package is missing).

Thanks for mention. fixed.

**Comment by Euklios** [comment]
I expose it the same way, as I do with the API-module, by including it as a dependency with the scope set to "provided" in the pom.xml of the video-plugin. Additionally added the dependency within the plugin.properties file.

So basically I have to either load the Entities and Repositories from the plugin and pass it to the core application, which sounds like a huge hassle to me.
Or I don't directly reference the entity and use a foreign key constraint in combination with a service, that loads the corresponding model based on a given id.
So basically something like:
```java
libraryService.getLibraryForId(video.getLibraryId())
```
instead of
```java
vide.getLibrary()
```

I also could just switch to something like iBatis/MyBatis, but I believe this will bring a different set of problems and maybe not even solve this one.

I will look into it later and post if I find something new

**Comment by hank-cp** [comment]
I spent so many time trying to manage hibernate entity graph, and gave it up finally. HIbernate is totally deaign for monolithic architecture, and never consider any runtime/dynamic support, as well as in future I guess.

In our projects using sbp, we are using jooq/[SimpleFlatMapper](https://github.com/arnaudroger/SimpleFlatMapper) combination, and manage model association mostly via graphql.

JPA/hibernate may be used in some very independent senario like batch processing.

**Comment by l-dietrich** [comment]
So I ran in to the same issue and solved it by adding the package of the "unmapped class" to the plugins starter class with an @EntityScan Annotation.
JPA therefore can find the class and map it.
e.g. 

```java
@EntityScan(basePackages = {"MAIN_PACKAGE_OF_THIS_PLUGIN","MAIN_PACKAGE_OF_ANOTHER_PLUGIN"})
@SpringBootApplication
public class DemoPluginsStarter {
...
```
You have to add the package of the actual plugin there as well otherwise only the other plugins package will be scanned.

**Comment by hank-cp** [comment]
> So I ran in to the same issue and solved it by adding the package of the "unmapped class" to the plugins starter class with an `@EntityScan` Annotation.
> JPA therefore can find the class and map it.
> e.g.
> 
> ```java
> @EntityScan(basePackages = {"MAIN_PACKAGE_OF_THIS_PLUGIN","MAIN_PACKAGE_OF_ANOTHER_PLUGIN"})
> @SpringBootApplication
> public class DemoPluginsStarter {
> ...
> ```
> 
> You have to add the package of the actual plugin there as well otherwise only the other plugins package will be scanned.

@l-dietrich In your case, there will be more than one EntityManager instances and not the case of `SharedDataSourceSpringBootstrap` from original question any more.

**Comment by hank-cp** [comment]
In the latest version `3.0-SNAPSHOT`, a `org.laxture:sbp-spring-boot-jpa-starter:3.0-SNAPSHOT` is introduced, trying to bring dynamic model mapping to a single `EntityManagerFactoryBean`.
